ports/esp32/boards: Add WAVESHARE_ESP32_S3_PICO board configuration.

EngineerWill opened the pull request:
Add WAVESHARE_ESP32_S3_PICO definition

projectgus commented:
@EngineerWill Sorry nobody ever got back to you about this. It needs a little bit of cleanup, but it's almost ready to merge.

Are you still interested in working on it? If so then I can do a quick review.

EngineerWill commented:
I'd be happy to handle it. What problems should I address? And sorry for the late reply.

projectgus commented on ports/esp32/boards/WAVESHARE_ESP32_S3_PICO/modules/pico_gpio.py:
This file can be replaced with a pins.csv file, for examples see other boards in the ports/esp32/boards directory.

projectgus commented on ports/esp32/boards/WAVESHARE_ESP32_S3_PICO/board.json:
```diff
@@ -0,0 +1,23 @@
+        "WiFi"
+    ],
+    "images": [
+        "generic_s3.jpg"
```
Please submit a PR adding an image of this board to https://github.com/micropython/micropython-media/
and then use the name here

projectgus commented on ports/esp32/boards/WAVESHARE_ESP32_S3_PICO/board.md:
```diff
@@ -0,0 +1 @@
+The following files are firmware for the WAVESHARE_ESP32_S3_PICO.
```
```suggestion
The following files are firmware for the Waveshare ESP32-S3-Pico.
```

projectgus commented on ports/esp32/boards/WAVESHARE_ESP32_S3_PICO/manifest.py:
This file can be removed once pico_gpio.py is replaced with pins.csv

projectgus commented on ports/esp32/boards/WAVESHARE_ESP32_S3_PICO/board.json:
```diff
@@ -0,0 +1,23 @@
+    "url": "https://www.waveshare.com/ESP32-S3-Pico.htm",
+    "vendor": "Waveshare",
+    "variants": {
+        "SPIRAM_OCT": "Support for SPIRAM"
```
Remove this unless there is a second variant of this board with Octal SPIRAM (I think there is not?)

projectgus commented on ports/esp32/boards/WAVESHARE_ESP32_S3_PICO/mpconfigboard.cmake:
This file structure has changed a bit since 2024. I think all you need for this file now is:

```cmake
include(boards/mpconfigboard_esp32s3_common.cmake)

list(APPEND SDKCONFIG_DEFAULTS
    boards/sdkconfig.flash_qio_80m)
```

projectgus commented on ports/esp32/boards/WAVESHARE_ESP32_S3_PICO/mpconfigboard.h:
```diff
@@ -0,0 +1,11 @@
+#ifndef MICROPY_HW_BOARD_NAME
+// Can be set by mpconfigboard.cmake.
+#define MICROPY_HW_BOARD_NAME               "WAVESHARE_ESP32_S3_PICO"
```
```suggestion
#define MICROPY_HW_BOARD_NAME               "Waveshare ESP32-S3-Pico"
```

projectgus commented on ports/esp32/boards/WAVESHARE_ESP32_S3_PICO/mpconfigboard.h:
```diff
@@ -0,0 +1,11 @@
+// Can be set by mpconfigboard.cmake.
+#define MICROPY_HW_BOARD_NAME               "WAVESHARE_ESP32_S3_PICO"
+#endif
+#define MICROPY_HW_MCU_NAME                 "ESP32S3"
```
```suggestion
#define MICROPY_HW_MCU_NAME                 "ESP32-S3"
```

projectgus commented on ports/esp32/boards/WAVESHARE_ESP32_S3_PICO/sdkconfig.board:
Can remove this file now (`sdkconfig.flash_qio_80m` handles the first settings, and MicroPython will now automatically use the 16MiB flash without the extra config.

projectgus reviewed (commented):
Hi @EngineerWill,

I've left some comments inline. First step will be to rebase this PR branch against the latest master branch (can find [some advice here](https://github.com/jimmo/git-and-micropython) if unfamiliar).

> And sorry for the late reply.

It took us 18 months to reply to you, a few days is pretty quick. 😅

projectgus commented:
@EngineerWill Are you still interested in updating this PR?

EngineerWill commented:
> @EngineerWill Are you still interested in updating this PR?

I’d be happy to handle it. I missed the notification, so I’ll take care of it today.

Commit:
esp32/boards: Add WAVESHARE_ESP32_S3_PICO board configuration.

Signed-off-by: eng33 <eng33@waveshare.com>

EngineerWill commented:
Hi @projectgus 
 
I've updated the PR based on your review feedback:
 
1. Replaced pico_gpio.py with pins.csv
2. Removed manifest.py and sdkconfig.board
3. Updated mpconfigboard.cmake to use the new structure (mpconfigboard_esp32s3_common.cmake)
4. Updated board.json (removed variants, added deploy_options)
5. Updated board.md and mpconfigboard.h with proper naming
  
I've also submitted a PR for the board image: https://github.com/micropython/micropython-media/pull/125
 
Please let me know if there's anything else that needs to be changed.

projectgus commented on ports/esp32/boards/WAVESHARE_ESP32_S3_PICO/pins.csv:
One thing I'm unsure about, if I look at the image of this board on the website:

![Waveshare board](https://www.waveshare.com/media/catalog/product/cache/1/image/800x800/9df78eab33525d08d6e5fb8d27136e95/e/s/esp32-s3-pico-2.jpg)

... the pin names don't seem to be `D0`, `D1`, `D2`, etc. If these labels aren't found anywhere then they can be removed from here (or you could add labels `GP11`, `GP12`, etc to match the labels - but I think most code will use the simpler `Pin(11)` or similar, and this works even without the entry in pins.csv).

Can still keep special pins for this board like `RGB_PIN`, though.

EngineerWill replied:
The reason for using D0, D1, D2 naming is that this board is designed to be compatible with Raspberry Pi Pico expansion boards. The pin layout references the Pico's GPIO naming convention (GP0-GP28).

Since the ESP32-S3's IO arrangement is different from the RP2040, users who want to migrate from Pico to this
board would need to remap pins. By providing the D0-D28 mapping in pins.csv, users can easily use existing Pico
expansion board code with minimal changes (e.g., from pico_gpio import D0 instead of manually figuring out which ESP32 pin corresponds to which Pico pin).

If you think this mapping is unnecessary, I can remove it and keep only special pins like RGB_PIN. What do you
recommend?

projectgus replied:
Oh I see, yes this makes sense if using an expansion board. I'm OK to keep it like it is, thanks for explaining.

projectgus reviewed (commented):
Thanks for the quick update, @EngineerWill! I have one question but otherwise this looks good to me.

projectgus reviewed (approved):
LGTM. Have built this PR locally to verify, but I don't have this exact board to test.

Thanks @EngineerWill for the quick updates!
